### PR TITLE
Add account activation

### PR DESCRIPTION
**Pull request by NguyenThu2808**

### CODE REVIEW

**Comment by NguyenThu2808** [comment]
ready

File: app/controllers/account_activations_controller.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class AccountActivationsController < ApplicationController
+  def edit
+    user = User.find_by email: params[:email]
```
Comment (mdmahmudurrahman) [inline app/controllers/account_activations_controller.rb]: If you don't find the user with the params[email] ??
You can do the same thing what you have done in the chapter 10 using filter.

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -3,12 +3,19 @@ def new; end
...
-      params[:session][:remember_me] == "1" ? remember(user) : forget(user)
-      redirect_back_or user
+    if user&.authenticate params[:session][:password]
+      if user.activated?
+        log_in user
+        params[:session][:remember_me] == "1" ? remember(user) : forget(user)
+        redirect_back_or user
+      else
+        message = t ".account_not_activated"
+        message += t ".check_email_link"
```
Comment (mdmahmudurrahman) [inline app/controllers/sessions_controller.rb]: Please combine line 13 and 12 into one.

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -1,33 +1,33 @@
...
+  def edit; end
+
   def create
     @user = User.new user_params
     if @user.save
-      log_in @user
-      flash[:success] = t ".welcome"
-      redirect_to @user
+      @user.send_activation_email
+      flash[:info] = "Please check your email to activate your account."
```
Comment (mdmahmudurrahman) [inline app/controllers/users_controller.rb]: I18n

File: app/mailers/user_mailer.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  def account_activation user
+    @user = user
+    mail to: user.email, subject: "Account acctivation"
+  end
+
+  # Subject can be set in your I18n file at config/locales/en.yml
+  # with the following lookup:
+  #
+  #   en.user_mailer.password_reset.subject
+  #
```
Comment (mdmahmudurrahman) [inline app/mailers/user_mailer.rb]: Please remove all of the comments

File: app/mailers/user_mailer.rb [hunk 5]
```diff
@@ -0,0 +1,23 @@
+class UserMailer < ApplicationMailer
+
+  # Subject can be set in your I18n file at config/locales/en.yml
+  # with the following lookup:
+  #
+  #   en.user_mailer.account_activation.subject
+  #
+  def account_activation user
+    @user = user
+    mail to: user.email, subject: "Account acctivation"
```
Comment (mdmahmudurrahman) [inline app/mailers/user_mailer.rb]: I18n

File: app/models/user.rb [hunk 6]
```diff
@@ -22,17 +23,38 @@ def new_token
     end
   end
 
+  def activate
+    update_attribute(:activated, true)
+    update_attribute(:activated_at, Time.zone.now)
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: 1. Please study about update_attributes and use it there.

File: app/models/user.rb [hunk 7]
```diff
@@ -22,17 +23,38 @@ def new_token
...
+
   def remember
     self.remember_token = User.new_token
     update_attributes remember_digest: User.digest(remember_token)
   end
 
-  def authenticated? remember_token
-    return false if remember_digest.nil?
-    BCrypt::Password.new(remember_digest).is_password? remember_token
+  def authenticated?(attribute, token)
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: Remove ()

File: app/models/user.rb [hunk 8]
```diff
@@ -22,17 +23,38 @@ def new_token
...
+    digest = send("#{attribute}_digest")
+    return false if digest.nil?
+    BCrypt::Password.new(digest).is_password?(token)
```
Comment (mdmahmudurrahman) [inline app/models/user.rb]: The same here

File: app/views/user_mailer/account_activation.html.erb [hunk 9]
```diff
@@ -0,0 +1,7 @@
+<h1><%= t ".sampleapp" %></h1>
+<p>Hi <%= @user.name %>,</p>
```
Comment (mdmahmudurrahman) [inline app/views/user_mailer/account_activation.html.erb]: I18n

File: app/views/user_mailer/account_activation.html.erb [hunk 10]
```diff
@@ -0,0 +1,7 @@
...
+<p>
+	<%= t ".welcome_click_link" %>
```
Comment (mdmahmudurrahman) [inline app/views/user_mailer/account_activation.html.erb]: Indentation

File: app/views/user_mailer/account_activation.text.erb [hunk 11]
```diff
@@ -0,0 +1,3 @@
+<%= t ".hi" %> <%= @user.name %>
+<%= t ".welcome_click_link" %>
+<%= edit_account_activation_url(@user.activation_token, email: @user.email) %>
```
Comment (mdmahmudurrahman) [inline app/views/user_mailer/account_activation.text.erb]: Remove ()

File: app/views/users/show.html.erb [hunk 12]
```diff
@@ -1,11 +1,11 @@
...
-    </section>
-  </aside>
+	<aside class="col-md-4">
+		<section class="user_info">
+			<h1>
+				<%= gravatar_for @user %>
+				<%= @user.name %>
+			</h1>
+		</section>
+	</aside>
```
Comment (mdmahmudurrahman) [inline app/views/users/show.html.erb]: Please correct the indentation

File: config/environments/development.rb [hunk 13]
```diff
@@ -31,7 +31,10 @@
   config.active_storage.service = :local
 
   # Don't care if the mailer can't send.
-  config.action_mailer.raise_delivery_errors = false
+  config.action_mailer.raise_delivery_errors = true
+  config.action_mailer.delivery_method = :test
+  host = "localhost:3000"
+  config.action_mailer.default_url_options = { host: host, protocol: "http" }
```
Comment (mdmahmudurrahman) [inline config/environments/development.rb]: Remove white space after { and before }

File: config/environments/test.rb [hunk 14]
```diff
@@ -37,6 +37,7 @@
   # The :test delivery method accumulates sent emails in the
   # ActionMailer::Base.deliveries array.
   config.action_mailer.delivery_method = :test
+  config.action_mailer.default_url_options = { host: "localhost:3000"}
```
Comment (mdmahmudurrahman) [inline config/environments/test.rb]: There same here

File: config/routes.rb [hunk 15]
```diff
@@ -9,4 +8,5 @@
   post "/login", to: "sessions#create"
   delete "/logout", to: "sessions#destroy"
   resources :users
+  resources :account_activations, only: [:edit]
```
Comment (mdmahmudurrahman) [inline config/routes.rb]: No need array for one action. 
`resources :account_activations, only: :edit
`

**Comment by mdmahmudurrahman** [comment]
commented

**Comment by NguyenThu2808** [comment]
ready

File: app/models/user.rb [hunk 16]
```diff
@@ -22,17 +23,38 @@ def new_token
     end
   end
 
+  def activate
+    update_attributes(:activated, true)
+    update_attributes(:activated_at, Time.zone.now)
```
Comment (huongnguyenmta) [inline app/models/user.rb]: gộp xử lý dòng 27, 28

**Comment by huongnguyenmta** [comment]
commented

File: app/controllers/users_controller.rb [hunk 17]
```diff
@@ -1,33 +1,33 @@
 class UsersController < ApplicationController
-  before_action :load_user, except: %i(index correct_user update destroy create)
-  before_action :logged_in_user, only: %i(index edit update)
+  before_action :load_user, except: %i(new create index correct_user)
```
Comment (huuphuc93) [inline app/controllers/users_controller.rb]: correct_user là method private nên k cần except đâu em

File: app/models/user.rb [hunk 18]
```diff
@@ -22,17 +23,38 @@ def new_token
     end
   end
 
+  def activate
+    update_attributes(:activated, true)
```
Comment (huuphuc93) [inline app/models/user.rb]: update_attributes(:activated, true) ; update_attributes(:activated_at, Time.zone.now) => update_attributes activated: true, activated_at: Time.zone.now

File: app/views/user_mailer/account_activation.html.erb [hunk 19]
```diff
@@ -0,0 +1,7 @@
+<h1><%= t ".sampleapp" %></h1>
+<p><%= t ".hi" %> <%= @user.name %>,</p>
+<p>
+	<%= t ".welcome_click_link" %>
```
Comment (huuphuc93) [inline app/views/user_mailer/account_activation.html.erb]: indent

**Comment by huuphuc93** [comment]
commented
comment nhỏ

**Comment by NguyenThu2808** [comment]
ready

**Comment by huongnguyenmta** [comment]
merged
Mợt được rồi nha gái ơi :*

**Pull request merged**